CorsixTH/Lua/announcer.lua

```diff
@@ -0,0 +1,157 @@
+  [AnnouncementPriority.High] = 2000,
+  [AnnouncementPriority.Normal] = 1000,
+  [AnnouncementPriority.Low] = 300
+}
+
+-- An announcement queue based on priority
+class "AnnouncementQueue"
+
+function AnnouncementQueue:AnnouncementQueue()
+  -- might want to use a single table for this
+  self.priorities = {
+    [AnnouncementPriority.Critical] = {},
+    [AnnouncementPriority.High] = {},
+    [AnnouncementPriority.Normal] = {},
+    [AnnouncementPriority.Low] = {}
+  }
+
+  self.count = 0
+end
+
+function AnnouncementQueue:push(priority, entry)
+  local entries = self.priorities[priority]
+  table.insert(entries, entry)
+  self.count = self.count + 1
+end
+
+function AnnouncementQueue:pop()
+  for priority,entries in pairs(self.priorities) do
+    if entries[1] ~= nil then
+      local entry = table.remove(entries, 1)
+      self.count = self.count - 1
+      return entry
+    end
+  end
+  
+  return nil
+end
+
+function AnnouncementQueue:isEmpty()
+  return self.count == 0
+end
+
+-- A single announcement
+class "AnnouncementEntry"
+
+function AnnouncementEntry:AnnouncementEntry()
+  self.name = nil -- filename to play
+  self.priority = default_announcement_priority
+  self.created_tick = nil -- when it has been created
+  self.decay_ticks = nil -- how long until the announcement isn't relevant anymore
+  self.played_callback = nil -- call me whenever the sound was played, ...
+  self.played_callback_delay = nil -- but not until delay has passed
+end
+
+class "Announcer"
+
+local Announcer = _G["Announcer"]
+
+function Announcer:Announcer(app)
+  self.app = app
+  self.entries = AnnouncementQueue()
+  self.playing = 0
+  self.ticks = 0
+  self.ticks_since_last_announcement = 0
+  
+  self:_setRandomAnnouncementTarget()
+end
+
+function Announcer:playAnnouncement(name, priority, decay_ticks, played_callback, played_callback_delay)
+  local priority = priority or default_announcement_priority
+  local created_tick = self.ticks
+  local decay_ticks = decay_ticks or default_announcement_decay_time[priority]
+  
+  local entry = AnnouncementEntry()
+  entry.name = name
+  entry.priority = priority
+  entry.created_tick = created_tick
+  entry.decay_ticks = decay_ticks
+  entry.played_callback = played_callback
+  entry.played_callback_delay = played_callback_delay
+
+  self.entries:push(priority, entry)
+  self:_debug("[+Announcement] " .. entry.name .. ", P: " .. entry.priority .. ", CT: " .. entry.created_tick .. ", DT: " .. entry.decay_ticks)
+end
+
+function Announcer:onTick()
+  if not self.app.world:isCurrentSpeed("Pause") then
+    local ticks_since_last_announcement = self.ticks_since_last_announcement
+    if ticks_since_last_announcement >= self.random_announcement_ticks_target then
+      self:playAnnouncement("rand*.wav", AnnouncementPriority.Low)
+      self:_setRandomAnnouncementTarget()
+    else
+      self.ticks_since_last_announcement = ticks_since_last_announcement + 1
+    end
+  end
+  
+  self.ticks = self.ticks + 1
```

Comment:
I think it would be better to use the game_date from world than tracking your own hours/in game ticks here.

Reply:
I've added a commit handling this issue.